File: client/src/modules/nexotype/schemas/knowledge_graph/source.schemas.ts
import { z } from 'zod';

// ==========================================
// Enums
// ==========================================

export const SourceTypeEnum = z.enum(['PubMed', 'Patent', 'ClinicalTrial']);

// ==========================================
// Options (for Select dropdowns)
// ==========================================

/** Source.source_type */
export const SOURCE_TYPE_OPTIONS = [
  { label: 'PubMed', value: 'PubMed' },
  { label: 'Patent', value: 'Patent' },
  { label: 'ClinicalTrial', value: 'ClinicalTrial' },
] as const;

// ==========================================
// Full Schema (GET responses)
// ==========================================

export const SourceSchema = z.object({
  id: z.number(),
  source_type: SourceTypeEnum,
  external_id: z.string().min(1).max(100),
  title: z.string().nullable(),
  authors: z.string().nullable(),
  journal: z.string().max(255).nullable(),
  publication_date: z.string().nullable(),
  url: z.string().max(500).nullable(),
  created_at: z.string(),
  updated_at: z.string().nullable(),
});

// ==========================================
// Input Schemas (POST/PUT)
// ==========================================

export const CreateSourceSchema = z.object({
  source_type: SourceTypeEnum,
  external_id: z.string().min(1).max(100),
  title: z.string().optional(),
  authors: z.string().optional(),
  journal: z.string().max(255).optional(),
  publication_date: z.string().optional(),
  url: z.string().max(500).optional(),
});

export const UpdateSourceSchema = z.object({
  source_type: SourceTypeEnum.optional(),
  external_id: z.string().min(1).max(100).optional(),
  title: z.string().optional(),
  authors: z.string().optional(),
  journal: z.string().max(255).optional(),
  publication_date: z.string().optional(),
  url: z.string().max(500).optional(),
});

// ==========================================
// Types
// ==========================================

export type Source = z.infer<typeof SourceSchema>;
export type CreateSource = z.infer<typeof CreateSourceSchema>;
export type UpdateSource = z.infer<typeof UpdateSourceSchema>;
export type SourceType = z.infer<typeof SourceTypeEnum>;

// ==========================================
// Response Types
// ==========================================

export type SourceResponse = {
  success: boolean;
  data?: Source;
  error?: string;
};

export type SourcesResponse = {
  success: boolean;
  data?: Source[];
  count?: number;
  error?: string;
};

File: server/apps/nexotype/schemas/knowledge_graph/source_schemas.py
from pydantic import BaseModel, ConfigDict
from datetime import date, datetime

# ==========================================
# Request Schemas
# ==========================================

class SourceCreate(BaseModel):
    """Schema for creating a new source"""
    source_type: str
    external_id: str
    title: str | None = None
    authors: str | None = None
    journal: str | None = None
    publication_date: date | None = None
    url: str | None = None


class SourceUpdate(BaseModel):
    """Schema for updating a source"""
    source_type: str = None
    external_id: str = None
    title: str | None = None
    authors: str | None = None
    journal: str | None = None
    publication_date: date | None = None
    url: str | None = None


# ==========================================
# Response Schemas
# ==========================================

class SourceDetail(BaseModel):
    """Schema for source details"""
    id: int
    source_type: str
    external_id: str
    title: str | None = None
    authors: str | None = None
    journal: str | None = None
    publication_date: date | None = None
    url: str | None = None
    created_at: datetime
    updated_at: datetime | None = None

    model_config = ConfigDict(from_attributes=True)


class SourceListResponse(BaseModel):
    """Response schema for listing sources"""
    success: bool
    data: list[SourceDetail] | None = None
    count: int = 0
    error: str | None = None


class SourceResponse(BaseModel):
    """Response schema for single source operations"""
    success: bool
    data: SourceDetail | None = None
    error: str | None = None


class MessageResponse(BaseModel):
    """Simple message response"""
    success: bool
    message: str | None = None
    error: str | None = None
